sklearn/metrics/_dist_metrics.pyx

```diff
@@ -1145,3 +1188,319 @@ cdef class PyFuncDistance(DistanceMetric):
 
 cdef inline double fmax(double a, double b) nogil:
     return max(a, b)
+
+
+######################################################################
+# Datasets Pair Classes
+cdef class DatasetsPair:
+    """Abstract class which wraps a pair of datasets (X, Y).
+
+    This class allows computing distances between two vectors (X_i, Y_j)
+    (rows of X and Y) at a time given the pair of their indices (i, j).
+
+    X and Y can be stored as np.ndarrays or CSR matrices in subclasses.
+
+    This class avoids the overhead of dispatching distance computations
+    to :class:`sklearn.metrics.DistanceMetric` based on the physical
+    representation of the vectors (sparse vs. dense). It makes use of
+    cython.final to remove the overhead of method calls' dispatch.
+    """
+
+    @classmethod
+    def get_for(cls,
+        X,
+        Y,
+        str metric="euclidean",
+        dict metric_kwargs=dict(),
+    ) -> DatasetsPair:
+        cdef:
+            DistanceMetric distance_metric = DistanceMetric.get_metric(metric,
+                                                                 **metric_kwargs)
+
+        # TODO: what's the best coercion for lists?
+        X = np.asarray(X) if isinstance(X, (tuple, list)) else X
```

Comment:
`check_array` would not do that?

Reply:
It indeed does.

Reply:
Ah, I remember why I was not doing it: potentially `check_array` is expensive and I just want to have a quick check the dtype of X and Y for non 32bitness.